### PR TITLE
Graduate CustomFasSyncInterval To Stable

### LINKED ISSUE #3095: Graduate `CustomFasSyncInterval` to Stable
I propose that we graduate the `CustomFasSyncInterval` feature to stable. It has been in beta since 1.25.0. 

Context:
https://development.agones.dev/site/docs/guides/feature-stages/#feature-gates

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature

> /kind hotfix

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3095

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by markmandel** [comment]
Looks like the agones-bot got stuck. If you rebase against `main` and push force over the top, CI should kick off.

**Comment by Kalaiselvi84** [comment]
I did rebase against main, still agones-bot is not running. Am I missing anything here?

**Comment by markmandel** [comment]
Well that's weird! I'll have to go look!

**Comment by markmandel** [comment]
Ah I see it. One second...

**Comment by Kalaiselvi84** [comment]
can I try creating a new PR?

**Comment by markmandel** [comment]
Nah, I broke some things, I just fixed them. Running now!

File: pkg/fleetautoscalers/controller.go [hunk 1]
```diff
@@ -122,28 +122,18 @@ func NewController(
 
 	autoscaler.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			if runtime.FeatureEnabled(runtime.FeatureCustomFasSyncInterval) {
-				c.addFasThread(obj.(*autoscalingv1.FleetAutoscaler), true)
```
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:126]: I see it! 😄 you deleted the wrong side of the condition here when cleaning up the code.

So we want to keep the side that is true for `runtime.FeatureEnabled`

File: pkg/fleetautoscalers/controller_test.go [hunk 2]
```diff
@@ -284,8 +282,6 @@ func TestControllerSyncFleetAutoscaler(t *testing.T) {
 	utilruntime.FeatureTestMutex.Lock()
 	defer utilruntime.FeatureTestMutex.Unlock()
```
[the remark this answers is not in the document]
Comment (Kalaiselvi84) [reply pkg/fleetautoscalers/controller_test.go]: I have removed it from all the test function in this file.

File: pkg/fleetautoscalers/controller.go (hunk 1 above)
Comment (Kalaiselvi84) [reply pkg/fleetautoscalers/controller.go:126]: I have substituted it with the correct statement. Both the `TestControllerEvents` function and `make test-go` have succeeded.

**Comment by markmandel** [comment]
`make gen-api-docs` should solve this issue I believe.

**Comment by Kalaiselvi84** [comment]
> `make gen-api-docs` should solve this issue I believe.

The execution of the make gen-api-docs command is encountering a failure at [this specific line](https://github.com/googleforgames/agones/blob/main/site/gen-api-docs.sh#L32)

**Comment by markmandel** [comment]
What is the error?

**Comment by Kalaiselvi84** [comment]
> What is the error?

[Here](https://paste.googleplex.com/5287185400987648) is the log for your reference

**Comment by markmandel** [comment]
Oh look at that - the tool assumes you have Go installed locally.

Install Go 😄

**Comment by Kalaiselvi84** [comment]
> Oh look at that - the tool assumes you have Go installed locally.
> 
> Install Go 😄

Go is already installed in my system 🙂
$ go version
go version go1.21-20230611-RC00 cl/539506128 +ee46f0b508 linux/amd64

**Comment by Kalaiselvi84** [comment]
Is it necessary for me to install Go version 1.15.13? Could the issue be related to the fact that I currently have Go version 1.21?

**Comment by markmandel** [comment]
Oh yeah, sorry - it should run in the container, so it should all just work. Huh, why do I not see it installing Go in the container - that's odd.

**Comment by markmandel** [comment]
Looking at your output, it's missing several steps. Have you changed the .sh file in any way? Running an old version somehow? running on a m2 mac?

This is my output when it runs against `main`

```
❯ make gen-api-docs
mkdir -p ~/.kube/
mkdir -p /home/markmandel/workspace/agones/build//.gocache
mkdir -p /home/markmandel/workspace/agones/build//.config/gcloud
mkdir -p ~/.config/helm
mkdir -p ~/.cache/helm
make ensure-image IMAGE_TAG=agones-build:09f8403ba1 BUILD_TARGET=build-build-image
make[1]: Entering directory '/home/markmandel/workspace/agones/build'
make[1]: Leaving directory '/home/markmandel/workspace/agones/build'
docker run -e FILE="/go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html" -e VERSION=1.33.0 --rm -i -v /home/markmandel/workspace/agones/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.config/helm:/root/.config/helm -v ~/.cache/helm:/root/.cache/helm -v /home/markmandel/workspace/agones:/go/src/agones.dev/agones -v /home/markmandel/workspace/agones/build//.gomod:/go/pkg/mod -v /home/markmandel/workspace/agones/build//.gocache:/root/.cache/go-build agones-build:09f8403ba1 bash -c "/go/src/agones.dev/agones/site/gen-api-docs.sh"
+ rm -rf /usr/local/go
+ GO_VERSION=1.15.13
+ cd /usr/local
+ wget -q https://dl.google.com/go/go1.15.13.linux-amd64.tar.gz
+ tar -xzf go1.15.13.linux-amd64.tar.gz
+ rm go1.15.13.linux-amd64.tar.gz
+ export GOPROXY=http://proxy.golang.org
+ GOPROXY=http://proxy.golang.org
+ echo 'using go proxy as a workaround for git.agache.org being down: http://proxy.golang.org'
+ cd /go/src/github.com/ahmetb/gen-crd-api-reference-docs
+ go mod edit --replace=agones.dev/agones@latest=../../../agones.dev/agones/
using go proxy as a workaround for git.agache.org being down: http://proxy.golang.org
+ go build
+ cp /go/src/agones.dev/agones/site/assets/templates/pkg.tpl ./template
+ FILE=/go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
+ VERSION=1.33.0
+ TITLE=/tmp/title.html
+ EXPIRY_DOC=/tmp/expiry.html
+ PUBLISH_DOC=/tmp/publish.html
+ RESULT=/tmp/agones_crd_api_reference.html
+ OLD=/tmp/old_docs.html
+ ./gen-crd-api-reference-docs --config ../../../agones.dev/agones/site/assets/templates/crd-doc-config.json --api-dir ../../../agones.dev/agones/pkg/apis/ --out-file /tmp/agones_crd_api_reference.html
I0629 22:07:16.517182     267 main.go:127] parsing go packages in directory ../../../agones.dev/agones/pkg/apis/
W0629 22:07:40.663014     267 parse.go:239] Ignoring child directory agones.dev/agones/pkg/apis/agones/v1/fuzz_test: No files for pkg "agones.dev/agones/pkg/apis/agones/v1/fuzz_test"
I0629 22:07:41.231152     267 main.go:229] using package=agones.dev/agones/pkg/apis/agones/v1
I0629 22:07:41.231163     267 main.go:229] using package=agones.dev/agones/pkg/apis/allocation/v1
I0629 22:07:41.231167     267 main.go:229] using package=agones.dev/agones/pkg/apis/autoscaling/v1
I0629 22:07:41.231169     267 main.go:229] using package=agones.dev/agones/pkg/apis/multicluster/v1
W0629 22:07:41.278711     267 main.go:423] not found external link source for type k8s.io/apimachinery/pkg/util/intstr.IntOrString
W0629 22:07:41.278823     267 main.go:423] not found external link source for type k8s.io/apimachinery/pkg/util/intstr.IntOrString
W0629 22:07:41.278907     267 main.go:423] not found external link source for type k8s.io/apimachinery/pkg/types.UID
W0629 22:07:41.279026     267 main.go:423] not found external link source for type k8s.io/apimachinery/pkg/types.UID
W0629 22:07:41.279941     267 main.go:423] not found external link source for type k8s.io/apimachinery/pkg/util/intstr.IntOrString
W0629 22:07:41.281033     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.281311     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.281718     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.282457     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.282865     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.283132     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.283804     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.284821     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
W0629 22:07:41.285176     267 main.go:423] not found external link source for type agones.dev/agones/pkg/apis.SchedulingStrategy
I0629 22:07:41.289015     267 main.go:165] written to /tmp/agones_crd_api_reference.html
+ awk '/\ feature\ publishVersion/{flag=1;next}/\ \/feature/{flag=0}flag' /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
+ awk '/\+\+\+/ {ok=1} /^$/ {ok=0} {if(ok){print $0}}' /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
+ printf '\n'
++ grep 'feature publishVersion=' /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
+ doc_version='{{% feature publishVersion="1.33.0" %}}'
+ echo '{{%' feature 'publishVersion="1.33.0"' '%}}'
+ publish='{{% feature publishVersion="1.33.0" %}}'
{{% feature publishVersion="1.33.0" %}}
+ expiry='{{% feature expiryVersion="1.33.0" %}}'
+ sed '/\ expiryVersion="1.33.0"/,/%\ \/feature/!d;/%\ \/feature/q' /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
+ sed '/\ publishVersion=/,/%\ \/feature/!d;/%\ \/feature/q' /go/src/agones.dev/agones/site/content/en/docs/Reference/agones_crd_api_reference.html
++ sort /tmp/agones_crd_api_reference.html
+ diff /dev/fd/63 /dev/fd/62
++ sort /tmp/old_docs.html
```

There should be a bunch more commands being run after `wget -q https://dl.google.com/go/go1.15.13.linux-amd64.tar.gz`

**Comment by Kalaiselvi84** [comment]
I haven't updated anything in the .sh file. The command fails to work on both my main branch and feature branch.

**Comment by markmandel** [comment]
I just ran the command on a clean machine (no build image, etc), and it all works correctly. So I'm not sure what to suggest 🤔

**Comment by Kalaiselvi84** [comment]
> I just ran the command on a clean machine (no build image, etc), and it all works correctly. So I'm not sure what to suggest 🤔

There was an issue in `load metadata for docker.io/library/alpine:3.14`. After executing the command `sudo systemctl restart docker` to restart the Docker service, it appears that the issue related to loading metadata for the alpine:3.14 image and making requests to the Docker registry (registry-1.docker.io) has been resolved.

**Comment by markmandel** [comment]
> There was an issue in `load metadata for docker.io/library/alpine:3.14`. After executing the command `sudo systemctl restart docker` to restart the Docker service, it appears that the issue related to loading metadata for the alpine:3.14 image and making requests to the Docker registry (registry-1.docker.io) has been resolved. `make gen-api-docs` command is now working fine.

Huh! Weird!

Glad you got it sorted!

File: pkg/fleetautoscalers/controller.go [hunk 3]
```diff
@@ -292,11 +282,6 @@ func (c *Controller) syncFleetAutoscaler(ctx context.Context, key string) error
 	}
 
 	if fas == nil {
-		// just in case we don't catch a delete event for some reason, use this as a
-		// failsafe to ensure we don't end up leaking goroutines.
-		if runtime.FeatureEnabled(runtime.FeatureCustomFasSyncInterval) {
-			return c.cleanFasThreads(key)
```
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:298]: `return c.cleanFasThreads(key) ` should have stayed in this block.

**Review by markmandel — COMMENTED** [review]
One small things to fix, otherwise, this looks good to go.

File: pkg/fleetautoscalers/controller.go (hunk 3 above)
Comment (Kalaiselvi84) [reply pkg/fleetautoscalers/controller.go:298]: I've updated with this correct statement

**Comment by Kalaiselvi84** [comment]
qq, I haven't added feature shortcodes in any of the files. Does everything look good?

File: site/content/en/docs/Guides/feature-stages.md [hunk 4]
```diff
@@ -26,7 +26,6 @@ The current set of `alpha` and `beta` feature gates:
 
 | Feature Name                                                                                                          | Gate                           | Default  | Stage   | Since  |
 |-----------------------------------------------------------------------------------------------------------------------|--------------------------------|----------|---------|--------|
-| [Custom resync period for FleetAutoscaler](https://github.com/googleforgames/agones/issues/1955)                      | `CustomFasSyncInterval`        | Enabled  | `Beta`  | 1.25.0 |
```
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:29]: Ooh! That was a good point - this table should be feature shortcoded.

File: site/content/en/docs/Reference/fleetautoscaler.md [hunk 5]
```diff
@@ -37,8 +37,6 @@ spec:
       # maximum fleet size that can be set by this FleetAutoscaler
       # required
       maxReplicas: 20
-  # [Stage:Beta]
```
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:40]: This update should also be feature shortcoded please.

**Review by markmandel — COMMENTED** [review]
Good point on the feature shortcodes 👍🏻

File: site/content/en/docs/Reference/fleetautoscaler.md (hunk 5 above)
Comment (Kalaiselvi84) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: I have a little doubt about this update. Is this correct?
  {{% feature expiryVersion="1.34.0" %}}
  # [Stage:Beta]
  # [FeatureFlag:CustomFasSyncInterval]
  {{% /feature %}}

Comment (markmandel) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: Thaaat might work? But I'm betting it won't.

If you run `make site-server` you can preview your changes locally and see what works and what doesn't.

Comment (Kalaiselvi84) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: Is the behavior shown in the screenshot at https://screenshot.googleplex.com/5nzmXVU6eacV8Kv expected?

There was an issue with `make site-server` command. This problem has been resolved by adding `git config --global --add safe.directory /go/src/agones.dev/agones` to the existing code.

Comment (markmandel) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: Yeah, that doesn't look right (it should be gone), you can also see the preview here: https://1b8d8e1-dot-preview-dot-agones-images.appspot.com/site/docs/reference/fleetautoscaler/

I expect you'll need to feature code the entire block.

Comment (Kalaiselvi84) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: I haven't observed any variation in my local environment, whether I utilize feature shortcodes or not. Even if I remove the entire block, I still encounter these lines:
```
# [Stage:Beta]
# [FeatureFlag:CustomFasSyncInterval]
```

Comment (markmandel) [reply site/content/en/docs/Reference/fleetautoscaler.md:40]: I think I wasn't clear - much like we do with markdown tables, we would have to duplicate the whole

```
```yaml
....
...
```

block.

Comment (Kalaiselvi84) [reply site/content/en/docs/Reference/fleetautoscaler.md]: I tried this too and it was not working in my local 😞 There might be some issue with my local env. committed the changes..

**Comment by markmandel** [comment]
https://cc537c5-dot-preview-dot-agones-images.appspot.com/site/docs/reference/fleetautoscaler/ looks good!

**Review by markmandel — APPROVED** [review]

**Pull request merged**